### ISSUE TITLE
Multi-GPU training performance issue

### ISSUE THREAD

**Issue by nbgbn** [StanfordVL/MinkowskiEngine#11]
Hi, 

I found that time consuming to train a model with multi gpus is totally the same as that with a single gpu.

And I did some experiments with demo [`examples/multigpu.py`](https://github.com/StanfordVL/MinkowskiEngine/blob/master/examples/multigpu.py):

```python3
for i in range(10): 
    optimizer.zero_grad()
    # Get new data
    inputs, labels = [], []
    for j in range(num_devices):
        coords, feat, label = data_loader()
        #print(coords)
        inputs.append(ME.SparseTensor(feat, coords=coords).to(devices[j]))
        labels.append(label.to(devices[j]))

    # The raw version of the parallel_apply
    t = time.time()
    replicas = parallel.replicate(net, devices)
    outputs = parallel.parallel_apply(replicas, inputs, devices=devices)
    # Time consumed
    print(time.time() - t, 's')
    
    #print(outputs)
    # Extract features from the sparse tensors to use a pytorch criterion
    out_features = [output.F for output in outputs]
    losses = parallel.parallel_apply(
        criterions, tuple(zip(out_features, labels)), devices=devices)
    loss = parallel.gather(losses, target_device, dim=0).mean()

    # Gradient
    loss.backward()
    optimizer.step()
```

When I run the code with 2 gpus, the output is as belows:
```shell
4.883297681808472 s
0.00571441650390625 s
0.005162477493286133 s
0.004988670349121094 s
0.004297733306884766 s
0.00394439697265625 s
0.004381656646728516 s
0.003942728042602539 s
0.004689216613769531 s
0.004220247268676758 s
```
While I run it with 8 gpus:
```shell
14.508163213729858 s
0.01927471160888672 s
0.017940759658813477 s
0.017798900604248047 s
0.017972707748413086 s
0.018915414810180664 s
0.020047664642333984 s
0.018342256546020508 s
0.01836872100830078 s
0.018011808395385742 s
```
I thought the time consumed with 2 gpus might approximate to that with 8 gpus because calculation should be paralleled. However the result is conflict to my expecation.

Is there a bug or other something?

**Comment by chrischoy** [comment]
The idea behind the data parallelism is to divide large batch into smaller ones and process them separately. If you create batch size 2 for 2 gpu and batch size 8 for 8 gpu, they will take the same time. I am not sure what data and model you used, but to make it less cpu bound, there has to be some adjustment to get the full speed up.

**Comment by chrischoy** [comment]
If you are using the naive version of the multi gpu, which uses 8 points with two conv layers, you would not get any speed up since processing 8 floating points is not what GPUs are optimized for...

**Comment by BuaaTang** [comment]
Thanks for your reply!

 I'm attempting to reproduce result from the Minkowski paper, and the model I used is just `MinkUNet34C`  that from [`examples/minkunet`](https://github.com/StanfordVL/MinkowskiEngine/blob/master/examples/minkunet.py). Actually,  batch size 8 for 8 gpu takes 4 times as long as batch size 2 for 2 gpu in my experiments. 

But I don't  understand you said "If you are using the naive version of the multi gpu, which uses 8 points with two conv layers, you would not get any speed up", is it regular that time consumed with using 8 gpus  more than 4 times than that with 2 gpus?

**Comment by chrischoy** [comment]
The current data loader loads really small data https://github.com/StanfordVL/MinkowskiEngine/blob/master/examples/common.py#L39 . Basically copying the result back to the gpu 0 would take more time than doing convolution on a small array.

**Comment by BuaaTang** [comment]
Is the opetation copying back to gpu0 time consuming?

I change [`examples/common.py#L39`](https://github.com/StanfordVL/MinkowskiEngine/blob/babcc93e02a5ac9e0fe93e2f92aef65f769b391e/examples/common.py#L39) to `data = [["   X   ", "  X X  ", " XXXXX "] for _ in range(100000)]`, and the data size may large enough, the output as below:

2 gpu:
```shell
6.26891303062439 s
2.123598098754883 s
2.0601632595062256 s
2.1894898414611816 s
1.9865307807922363 s
2.0760393142700195 s
2.014101028442383 s
2.026172637939453 s
2.0056533813476562 s
2.024648666381836 s
````

8 gpu:
```shell
20.48965311050415 s
6.981744289398193 s
6.968326568603516 s
6.913405418395996 s
6.957228183746338 s
7.025070428848267 s
6.877858638763428 s
7.202362775802612 s
7.0635621547698975 s
7.157938003540039 s
```
It seems there is no significant advantage to use multi-gpu training

**Comment by chrischoy** [comment]
What is the time for one gpu experiment and how many threads did you use? It depends also a lot on CPU as well.

**Comment by BuaaTang** [comment]
I‘m in a docker enviriment, and I don't know if it matters. Typically I run code with 2080ti and allocate 8 CPU cores, while the CPU usage is less than 50%. Here is result I test code with one GPU.
```shell
0.7296576499938965 s
0.7148511409759521 s
0.7068219184875488 s
0.718878984451294 s
0.7239933013916016 s
0.692812442779541 s
0.7242536544799805 s
0.707554817199707 s
0.6859543323516846 s
0.679347038269043 s
```

**Comment by BuaaTang** [comment]
It's difficult to create a fair enviroment to compare time took with different amount of gpus for me. But it seems using 2 gpus does speed up while over 2 gpus helps little to reduce time.

**Comment by chrischoy** [comment]
I see, but it seems like it is a bit problematic. I'll look into this further.

**Comment by wbhu** [comment]
Is it possible to use the native torch.nn.parallel.DistributedDataParallel to wrap the model?

**Comment by chrischoy** [comment]
This issue will be addressed in the next version with a new multi-thread safe hash table.

**Comment by chrischoy** [comment]
The issue has been resolved.

Multi GPU with batch size 2 x 2 GPU takes about 0.8sec
Single GPU with batch size 4 x 1 GPU takes about 0.77 sec.

There is some threading overhead, but the speed up is close to 2x.

```
(py3-align-1.3) ➜ chrischoy@napoli116 ~/projects/MinkowskiEngine git:(master) ✗ python -m examples.multigpu2 --max_ngpu 2 --batch_size 2                                                                                                                                                                                                                                                                              {19-12-15 18:25}
Testing  2  GPUs
Reading PLY: [========================================] 100%
/cvgl2/u/chrischoy/anaconda3/envs/py3-align-1.3/lib/python3.7/site-packages/torch/nn/parallel/_functions.py:61: UserWarning: Was asked to gather along dimension 0, but all input tensors were scalars; will instead unsqueeze and return a vector.                                                                                                                                                                                   
  warnings.warn('Was asked to gather along dimension 0, but all '
Iteration:  0 , Loss:  4.152383327484131 , Time:  1.6452147960662842
Iteration:  1 , Loss:  0.5146856904029846 , Time:  0.8751227855682373
Iteration:  2 , Loss:  0.13223637640476227 , Time:  0.840630054473877
Iteration:  3 , Loss:  0.05618448927998543 , Time:  0.8486003875732422
Iteration:  4 , Loss:  0.03924577683210373 , Time:  0.8670516014099121
Iteration:  5 , Loss:  0.03053881973028183 , Time:  0.8953015804290771
Iteration:  6 , Loss:  0.025215284898877144 , Time:  0.8200812339782715
Iteration:  7 , Loss:  0.021495893597602844 , Time:  0.8042786121368408
Iteration:  8 , Loss:  0.018669743090867996 , Time:  0.8148353099822998
Iteration:  9 , Loss:  0.016512056812644005 , Time:  0.8008191585540771
(py3-align-1.3) ➜ chrischoy@napoli116 ~/projects/MinkowskiEngine git:(master) ✗ python -m examples.multigpu2 --max_ngpu 1 --batch_size 4                                                                                                                                                                                                                                                                              {19-12-15 18:25}
Testing  1  GPUs
Reading PLY: [========================================] 100%
/cvgl2/u/chrischoy/anaconda3/envs/py3-align-1.3/lib/python3.7/site-packages/torch/nn/parallel/_functions.py:61: UserWarning: Was asked to gather along dimension 0, but all input tensors were scalars; will instead unsqueeze and return a vector.                                                                                                                                                                                   
  warnings.warn('Was asked to gather along dimension 0, but all '
Iteration:  0 , Loss:  2.9249300956726074 , Time:  1.2581493854522705
Iteration:  1 , Loss:  0.1864200383424759 , Time:  0.8700690269470215
Iteration:  2 , Loss:  0.06147165223956108 , Time:  0.7900640964508057
Iteration:  3 , Loss:  0.03992412984371185 , Time:  0.7602770328521729
Iteration:  4 , Loss:  0.03096628561615944 , Time:  0.8032622337341309
Iteration:  5 , Loss:  0.02524588443338871 , Time:  0.8056776523590088
Iteration:  6 , Loss:  0.02131427824497223 , Time:  0.7965397834777832
Iteration:  7 , Loss:  0.018375979736447334 , Time:  0.7677557468414307
Iteration:  8 , Loss:  0.016162533313035965 , Time:  0.8164584636688232
Iteration:  9 , Loss:  0.014465886168181896 , Time:  0.7711424827575684
```

**Comment by chrischoy** [comment]
Please checkout the finalized multi-gpu documentation with timing https://stanfordvl.github.io/MinkowskiEngine/demo/multigpu.html

**Comment by chrischoy** [comment]
On TitanXP, we have this speed up.

| Number of GPUs | Batch size per GPU | Time per iteration | Speedup |
|:--------------:|:------------------:|:------------------:|:-------:|
| 1 GPU          | 8                  | 1.611 s            | x1      |
| 2 GPU          | 4                  | 0.916 s            | x1.76   |
| 4 GPU          | 2                  | 0.689 s            | x2.34   |
